### PR TITLE
fix(stream): add group key into stream group topN's stream key

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Like `overWindow` and join, when update the partition key, we can not distinguish which one happend first in two different partition.

```
dev=> CREATE TABLE t(a int, b int, k int primary key);
explain create materialized view mv as select a,b,k from (
    select a,b,k, rank() over (partition by a order by b) as r from t
) tt where r < 10;
CREATE_TABLE
                                            QUERY PLAN                                            
--------------------------------------------------------------------------------------------------
 StreamMaterialize { columns: [a, b, k], stream_key: [k], pk_columns: [k], pk_conflict: NoCheck }
 └─StreamExchange { dist: HashShard(t.k) }
   └─StreamGroupTopN { order: [t.b ASC], limit: 9, offset: 0, with_ties: true, group_key: [t.a] }
     └─StreamExchange { dist: HashShard(t.a) }
       └─StreamTableScan { table: t, columns: [a, b, k] }
```

after the pr
```
dev=> CREATE TABLE t(a int, b int, k int primary key);
explain create materialized view mv as select a,b,k from (
    select a,b,k, rank() over (partition by a order by b) as r from t
) tt where r < 10;
CREATE_TABLE
                                               QUERY PLAN                                               
--------------------------------------------------------------------------------------------------------
 StreamMaterialize { columns: [a, b, k], stream_key: [k, a], pk_columns: [k, a], pk_conflict: NoCheck }
 └─StreamGroupTopN { order: [t.b ASC], limit: 9, offset: 0, with_ties: true, group_key: [t.a] }
   └─StreamExchange { dist: HashShard(t.a) }
     └─StreamTableScan { table: t, columns: [a, b, k] }
(4 rows)
```

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by stdrc — APPROVED** [review]

File: src/frontend/planner_test/tests/testdata/output/nexmark.yaml [hunk 1]
```diff
@@ -1567,24 +1567,19 @@
...
   stream_plan: |-
-    StreamMaterialize { columns: [auction, bidder, price, channel, url, date_time, extra, bid._row_id(hidden)], stream_key: [bid._row_id], pk_columns: [bid._row_id], pk_conflict: NoCheck }
-    └─StreamExchange { dist: HashShard(bid._row_id) }
-      └─StreamGroupTopN [append_only] { order: [bid.date_time DESC], limit: 1, offset: 0, with_ties: true, group_key: [bid.bidder, bid.auction] }
-        └─StreamExchange { dist: HashShard(bid.bidder, bid.auction) }
-          └─StreamTableScan { table: bid, columns: [bid.auction, bid.bidder, bid.price, bid.channel, bid.url, bid.date_time, bid.extra, bid._row_id], pk: [bid._row_id], dist: UpstreamHashShard(bid._row_id) }
+    StreamMaterialize { columns: [auction, bidder, price, channel, url, date_time, extra, bid._row_id(hidden)], stream_key: [bid._row_id, bidder, auction], pk_columns: [bid._row_id, bidder, auction], pk_conflict: NoCheck }
+    └─StreamGroupTopN [append_only] { order: [bid.date_time DESC], limit: 1, offset: 0, with_ties: true, group_key: [bid.bidder, bid.auction] }
```
Comment (fuyufjh) [inline src/frontend/planner_test/tests/testdata/output/nexmark.yaml:1571]: Shall we ignore that cases of append-only?

File: src/frontend/planner_test/tests/testdata/output/nexmark.yaml [hunk 2]
```diff
@@ -2354,7 +2339,7 @@
             └─BatchExchange { order: [], dist: HashShard(bid.auction) }
               └─BatchScan { table: bid, columns: [bid.auction], distribution: SomeShard }
   stream_plan: |-
-    StreamMaterialize { columns: [auction_id, auction_item_name, bid_count], stream_key: [auction_id, auction_item_name], pk_columns: [bid_count, auction_id, auction_item_name], pk_conflict: NoCheck }
+    StreamMaterialize { columns: [auction_id, auction_item_name, bid_count], stream_key: [], pk_columns: [bid_count], pk_conflict: NoCheck }
```
Comment (fuyufjh) [inline src/frontend/planner_test/tests/testdata/output/nexmark.yaml:2342]: Why does the `stream_key` become empty here?

**Comment by st1page** [comment]
It is because of the 2-phase agg, the globalAgg's stream key can not include vnode column

https://github.com/risingwavelabs/risingwave/pull/11587/commits/50b16ddfc6cfbddd36e0084327805b86413d3490
PTAL again @fuyufjh @chenzl25

File: src/frontend/planner_test/tests/testdata/output/nexmark.yaml (hunk 1 above)
Comment (st1page) [reply src/frontend/planner_test/tests/testdata/output/nexmark.yaml:1571]: Yes, but we can not get the append-only property in logical phase, but the derivation is in logical phase

**Review by chenzl25 — APPROVED** [review]
LGTM. Do any other operators have this problem? For example, OverWindow.

File: src/stream/src/from_proto/group_top_n.rs [hunk 3]
```diff
@@ -68,7 +68,7 @@ impl<const APPEND_ONLY: bool> ExecutorBuilder for GroupTopNExecutorBuilder<APPEN
                 &node.group_key.iter().map(|idx| *idx as usize).collect_vec()
             );
         } else {
-            assert_eq!(&params.pk_indices, input.pk_indices());
+            // assert_eq!(&params.pk_indices, input.pk_indices());
```
Comment (chenzl25) [inline src/stream/src/from_proto/group_top_n.rs:71]: Remove it?

**Comment by st1page** [comment]
> LGTM. Do any other operators have this problem? For example, OverWindow.

OverWindow solve this problem in development, just reconfirm @stdrc 

Agg is special that use the group key as pk. GroupTop1 is the same reason that there is no the same problem

**Comment by stdrc** [comment]
> OverWindow solve this problem in development, just reconfirm @stdrc

Yes, OverWindow luckily solved this during development, because of early encountering the bug.

**Pull request merged**